### PR TITLE
feat: add some s2 geo functions

### PR DESCRIPTION
**Pull request by sunng87**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This patch introduces some new s2 geo encoding functions.

More information: http://s2geometry.io/

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/common/function/src/scalars/geo/s2.rs [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+            signatures.push(TypeSignature::Exact(vec![
+                cell_type.clone(),
+                level_type.clone(),
+            ]));
+        }
+    }
+    Signature::one_of(signatures, Volatility::Stable)
+}
+
+fn cell_from_value(v: Value) -> Result<Option<CellID>> {
```
Comment (v0y4g3r) [inline src/common/function/src/scalars/geo/s2.rs:255]:
```suggestion
fn cell_from_value(v: Value) -> Option<CellID> {
```

**Review by v0y4g3r — APPROVED** [review]
Rest LGTM

File: src/common/function/src/scalars/geo.rs [hunk 2]
```diff
@@ -55,5 +54,11 @@ impl GeoFunctions {
         registry.register(Arc::new(h3::H3GridDiskDistances));
         registry.register(Arc::new(h3::H3GridDistance));
         registry.register(Arc::new(h3::H3GridPathCells));
+
+        //s2
```
Comment (waynexia) [inline src/common/function/src/scalars/geo.rs:58]:
```suggestion
        // s2
```

**Comment by adminpass** [comment]
Please support these functions:

[ST_Contains](https://help.aliyun.com/document_detail/2362789.html?spm=a2c4g.2362786.0.0.58b27765WjvyQy#section-4hs-xy5-r4z)
[ST_Within](https://help.aliyun.com/document_detail/2362789.html?spm=a2c4g.2362786.0.0.58b27765WjvyQy#section-19o-pvc-n5r)
[ST_Intersects](https://help.aliyun.com/document_detail/2362789.html?spm=a2c4g.2362786.0.0.58b27765WjvyQy#section-5r5-z9e-63n)

[ST_Area](https://help.aliyun.com/document_detail/2362791.html?spm=a2c4g.2362790.0.0.e0c33e8eVJnOJw#section-qjh-lbm-m8t)
[ST_Distance](https://help.aliyun.com/document_detail/2362791.html?spm=a2c4g.2362790.0.0.e0c33e8eVJnOJw#section-iln-m2u-c7z)
[ST_DistanceSphere](https://help.aliyun.com/document_detail/2362791.html?spm=a2c4g.2362790.0.0.e0c33e8eVJnOJw#section-vpj-kbv-xsl)

**Comment by sunng87** [comment]
@adminpass Thank you for letting me know! I wonder if you have a use-case with geospatial functions in GreptimeDB or other time-series database? It will be helpful if you can describe the scenario so we can prioritize.

**Comment by adminpass** [comment]
> @adminpass Thank you for letting me know! I wonder if you have a use-case with geospatial functions in GreptimeDB or other time-series database? It will be helpful if you can describe the scenario so we can prioritize.

For example:
1. Query coordinate data within the region
2. Query data with a specified coordinate radius of 10km

**Pull request merged**